plugins/auto-sizes/hooks.php

```diff
@@ -216,3 +216,37 @@ function auto_sizes_improve_image_sizes_attributes( string $content ): string {
 }
 // Run filter prior to auto sizes "auto_sizes_update_content_img_tag" filter.
 add_filter( 'wp_content_img_tag', 'auto_sizes_improve_image_sizes_attributes', 9 );
+
+/**
+ * Filter the sizes attribute for picture element.
+ *
+ * @since n.e.x.t
+ *
+ * @param string $content The block content about to be rendered.
+ * @return string The updated block content.
+ */
+function auto_sizes_improve_image_sizes_attributes_for_picture_element( string $content ): string {
+	$processor = new WP_HTML_Tag_Processor( $content );
+	if ( ! $processor->next_tag( array( 'tag_name' => 'picture' ) ) ) {
+		return $content;
+	}
+
+	$has_image = $processor->next_tag( array( 'tag_name' => 'img' ) );
+	$sizes     = null;
+	// Only update the markup if an image is found.
+	if ( $has_image ) {
+		$sizes = $processor->get_attribute( 'sizes' );
+	}
+
+	if ( ! $sizes ) {
+		return $content;
+	}
+
+	$processor = new WP_HTML_Tag_Processor( $content );
+	while ( $processor->next_tag( array( 'tag_name' => 'source' ) ) ) {
+		$processor->set_attribute( 'sizes', $sizes );
```

Comment:
What if there is already a `sizes` attribute on this `source` tag? Shouldn't it do nothing in that case?

```suggestion
		if ( ! is_string( $processor->get_attribute( 'sizes' ) ) ) {
			$processor->set_attribute( 'sizes', $sizes );
		}
```

Reply:
The `sizes` attribute is already set https://github.com/WordPress/performance/blob/trunk/plugins/webp-uploads/picture-element.php#L130 for `source` tags. We are updating it to be accurate based on the new API changes.

We take the correct `sizes` value from the `img` tag and update it in the `source` tags.